update: Try harder to fetch and merge subdataset revision

kyleam opened the pull request:
The motivation for this series stemmed from a scenario where, on the remote end, the submodule revision is the commit pointed to by another branch, say "other".  `update(..., merge=True)` called while on master in the local repo may choose to merge in something different, most likely origin/master, which let's say hasn't diverged from the local master (though either way it's problematic).  Then it saves that state as the submodule revision in the parent, bringing the submodule revision back to where it was before the remote update.

<details>
<summary>In code</summary>

```sh
cd "$(mktemp -d --tmpdir dl-XXXXXXX)"

datalad create rem
datalad create rem/sub
datalad -C rem save -r

datalad install -r -s rem ds

(
    cd rem/sub
    git checkout -b other
    echo foo >foo
)
datalad -C rem save -r

datalad -C ds update -r --merge
```

</details>

A very similar case can happen when the remote submodule has a detached head, a common occurrence when working with `git submodule` directly.  In this case, an additional issue is that the submodule revision might not even be fetched from the remote if the revision isn't contained in a ref that is pulled down on the fetch.

---

```
  [ 1/12] ENH: repo: Teach get_tracking_branch() to optionally ignore "."
  [ 2/12] BF: update: Don't consider "." as an upstream remote
  [ 3/12] BF: update: Query intended branch.*.remote variable
  [ 4/12] RF: update: Cosmetic changes to _update_repo() helper
  [ 5/12] RF: update: Avoid overloading a variable for clarity
  [ 6/12] OPT: update: Avoid repeatedly accessing Dataset.path in a loop
  [ 7/12] RF: update: Break apart _update_repo() helper
  [ 8/12] ENH: update: Use same merge approach for plain and annex repos
  [ 9/12] ENH: update: Avoid calling 'git pull' to merge changes
  [10/12] ENH: update: Don't set merge target to a guess that doesn't exist
  [11/12] ENH: update: Try to fetch submodule commit explicitly if needed
  [12/12] ENH: update: Prefer merge of submodule revision to a ref without it
```

Patch 1-3 are unrelated bugs that I noticed in the code I was touching for the fix.  These (or something like them) should be split off into a separate PR if the more substantial changes in this PR don't make the cut.

Patch 4-7 are cleanup/prep patches, with no intended change in behavior.

Patch 8-12 are the proposed fix, moving to a more targeted `git merge` to bring in changes rather using `git pull <remote> [<configured upstream or guess>]` in git repos or using `git annex sync <remote>`  in annex repos.  Note that, when we're on an adjusted branch, we keep using `git annex sync <remote>` as we did before.  It's not that this is correct or leads to good behavior; it's just that I don't see an alternative because using submodules and adjusted branches together is fundamentally problematic, and this PR isn't trying to address that.

mih commented on datalad/support/annexrepo.py:
```diff
@@ -506,7 +506,8 @@ def get_corresponding_branch(self, branch=None):
 
-    def get_tracking_branch(self, branch=None, corresponding=True):
+    def get_tracking_branch(self, branch=None, remote_only=False,
+                            corresponding=True):
```
Maybe swap order, such that `corresponding` remains the 2nd? I am torn whether backward-friendly keyword-less access is superior to GitRepo homogeneity or vice versa...

kyleam replied:
Right, that's the main trade-off.  I originally went the other way, but then decided it'd be better to keep the parent-child API between GitRepo/AnnexRepo in agreement.  Beyond a general preference, I think it particularly makes sense in this case because otherwise we introduce a hard-to-detect case where `ds.repo.get_tracking_branch(x, y)` does different things depending on whether you're in a plain repo or an annex repo.  I didn't spot any actual callers that use `corresponding` (positional or keyword) in the main code base or any of our extensions.

kyleam replied:
An alternative that results in a clearer failure mode:  We could make everything after `branch` keyword-only.

mih replied:
After another night of sleep, I think your solution is the right one.

kyleam commented on datalad/distribution/tests/test_update.py:
```diff
@@ -442,3 +446,165 @@ def test_multiway_merge(path):
+            ds_clone.update(recursive=True, on_failure="ignore"),
+            status="impossible",
+            path=ds_clone_s1.path,
+            action="update")
```
I ran these tests locally on top of gh-4087 (6e6210082), and this assertion fails.  A commit here adds a `fetch()` call that catches a `GitCommandError`, but that will be a `CommandError` following gh-4087.  So, assuming that gh-4087 will be merged first, this PR will need to be adjusted.

yarikoptic commented:
looks good to me.  Should "do-not-merge" label be replaced with "merge-if-ok"? ;-)

kyleam commented:
@yarikoptic Thanks for taking a look.

> Should "do-not-merge" label be replaced with "merge-if-ok"? ;-)

I marked it as such after [noting][0] the conflict with the then-in-flight gh-4087.  (Sorry, I should have probably made a top-level comment that linked the label to the conflict.)  Anyway, that's merged now, so I'll update this PR.

[0]: https://github.com/datalad/datalad/pull/4118#discussion_r377764247

kyleam commented:
I've updated this to resolve the conflicts with gh-4087 as well as trivial import conflicts.   There was also test bug where an assertion was meant to be about the current command but was actually being made about the previous one :/

<details>
<summary>range-diff</summary>

```
 1:  095bc7d5a1 =  1:  3485886ddb ENH: repo: Teach get_tracking_branch() to optionally ignore "."
 2:  94763387c4 =  2:  6ee5b663a3 BF: update: Don't consider "." as an upstream remote
 3:  b71b6de389 =  3:  422b904806 BF: update: Query intended branch.*.remote variable
 4:  36411bd612 =  4:  996cddb958 RF: update: Cosmetic changes to _update_repo() helper
 5:  049e036878 =  5:  e33375f216 RF: update: Avoid overloading a variable for clarity
 6:  d88b6fb9c4 =  6:  a08d8878d4 OPT: update: Avoid repeatedly accessing Dataset.path in a loop
 7:  871fa9826c =  7:  d0698f56a0 RF: update: Break apart _update_repo() helper
 8:  37cfaa7701 =  8:  1d8dac50d8 ENH: update: Use same merge approach for plain and annex repos
 9:  7523a854f4 =  9:  e7dc6d2948 ENH: update: Avoid calling 'git pull' to merge changes
10:  2368b7377b = 10:  0ea6cf46da ENH: update: Don't set merge target to a guess that doesn't exist
11:  cf1987c7c8 ! 11:  f7099dc1d5 ENH: update: Try to fetch submodule commit explicitly if needed
    @@ datalad/distribution/tests/test_update.py
      
      """
      
    --
    --
     +import logging
      import os
      import os.path as op
    - from os.path import join as opj, exists
    + from os.path import (
     @@
    - from datalad.utils import knows_annex
    - from datalad.utils import rmtree
    - from datalad.utils import chpwd
    -+from datalad.utils import Path
    -+from datalad.utils import swallow_logs
    +     knows_annex,
    +     rmtree,
    +     chpwd,
    ++    Path,
    ++    swallow_logs,
    + )
      from datalad.support.gitrepo import (
          GitRepo,
    -     GitCommandError,
      )
      from datalad.support.annexrepo import AnnexRepo
    - 
    -+from nose import SkipTest
    - from nose.tools import eq_, assert_false, assert_is_instance, ok_
    - from datalad.tests.utils import with_tempfile, assert_in, \
    -     with_testrepos, assert_not_in
    - from datalad.tests.utils import create_tree
    - from datalad.tests.utils import ok_file_has_content
    - from datalad.tests.utils import ok_clean_git
    -+from datalad.tests.utils import assert_repo_status
    - from datalad.tests.utils import assert_status
    - from datalad.tests.utils import assert_result_count
    - from datalad.tests.utils import assert_in_results
    -+from datalad.tests.utils import assert_not_in_results
    - from datalad.tests.utils import slow
    - from datalad.tests.utils import known_failure_windows
    - 
    ++from datalad.support.exceptions import CommandError
    + from datalad.tests.utils import (
    +     with_tempfile,
    +     assert_in,
    +@@
    +     create_tree,
    +     ok_file_has_content,
    +     ok_clean_git,
    ++    assert_repo_status,
    +     assert_status,
    +     assert_result_count,
    +     assert_in_results,
    ++    assert_not_in_results,
    ++    SkipTest,
    +     slow,
    +     known_failure_windows,
    + )
     @@ datalad/distribution/tests/test_update.py: def test_multiway_merge(path):
          assert_status('ok', ds.update())
          # ATM we do not support multi-way merges
    @@ datalad/distribution/tests/test_update.py: def test_multiway_merge(path):
     +    # This is the default, but just in case:
     +    ds_src_s1.repo.config.set("uploadpack.allowAnySHA1InWant", "false",
     +                              where="local")
    -+    res = ds_clone.update(merge=True, recursive=True, on_failure="ignore")
    -+    if on_adjusted:
    -+        assert_not_in_results(res, status="error")
    -+    else:
    -+        # The fetch with the explicit ref fails because it isn't advertised.
    -+        assert_in_results(
    -+            res,
    -+            status="impossible",
    -+            path=ds_clone_s1.path,
    -+            action="update")
    ++    # Note: Before the fetch() rewrite in 2e179b5588 (gh-4087), this would have
    ++    # resulted in an "impossible" result.
    ++    try:
    ++        ds_clone.update(merge=True, recursive=True)
    ++    except CommandError:
    ++        # The fetch with the explicit ref fails because it isn't advertised,
    ++        # unless we're on an adjusted branch, in which case `git annex sync`
    ++        # will be used.
    ++        if on_adjusted:
    ++            raise AssertionError(
    ++                "Unexpected failure of update() call on an adjusted branch")
     +
     +    # Back to the detached head.
     +    ds_src_s1.repo.checkout("HEAD@{1}")
    @@ datalad/distribution/tests/test_update.py: def test_multiway_merge(path):
     +    ds_clone_s1.repo.call_git(["branch", "--unset-upstream"])
     +    ds_clone_s1.config.reload(force=True)
     +    ds_clone_s1.repo.call_git(["remote", "add", "other", ds_src_s1.path])
    ++    res = ds_clone.update(recursive=True, on_failure="ignore")
     +    if on_adjusted:
     +        assert_not_in_results(res, status="error")
     +    else:
    @@ datalad/distribution/tests/test_update.py: def test_multiway_merge(path):
     +        # it does if the revision wasn't brought down in the `fetch(all_=True)`
     +        # call.
     +        assert_in_results(
    -+            ds_clone.update(recursive=True, on_failure="ignore"),
    ++            res,
     +            status="impossible",
     +            path=ds_clone_s1.path,
     +            action="update")
12:  836ae37fb2 = 12:  176ab1da55 ENH: update: Prefer merge of submodule revision to a ref without it
```

</details>

kyleam commented:
Marked as "do not merge" again.  It just crossed my mind that I updated for gh-4087 in the wrong spot (tests instead of `update` itself).

kyleam commented:
> Marked as "do not merge" again. It just crossed my mind that I updated for [gh-4087](https://github.com/datalad/datalad/pull/4087) in the wrong spot (tests instead of `update` itself).

Updated.  Collectively, the changes since Yarik commented:

*  As of gh-4087, `fetch` will raise a `CommandError` rather than a `GitCommandError`. Update the new `fetch` call in patch 11 to that catch a `CommandError`.

* Deal with the import conflicts from the recently merged "nose imports" PR.

* Fix last assertion in `check_merge_subdataset_detached` that, for the "on adjusted branch" case, made its assertion with the result record from wrong `update` command (the previous one).

* Make the new `fetch` call in patch 11 use `--recurse-submodules=no`, as the `fetch` call above it already does.  That existing call already has a comment that explains why it's necessary.  To expand on that, it makes sense because we're already iterating over submodules when `recursive=True`, so we want to avoid git doing repeated work (in the best case scenario) or the fetch call failing (in the worst case scenario, which is likely because submodule.c hardcodes "origin").

* Fix a missing period between sentences in a log message.

`update` should also probably be refreshed to use the new `fetch_` method added in gh-4087.  With that, it could yield a result record for each fetched remote.  But I think that should be handled separately.  (I'll open an issue.)

<details>
<summary>range-diff</summary>

```
 1:  095bc7d5a1 =  1:  3485886ddb ENH: repo: Teach get_tracking_branch() to optionally ignore "."
 2:  94763387c4 =  2:  6ee5b663a3 BF: update: Don't consider "." as an upstream remote
 3:  b71b6de389 =  3:  422b904806 BF: update: Query intended branch.*.remote variable
 4:  36411bd612 =  4:  996cddb958 RF: update: Cosmetic changes to _update_repo() helper
 5:  049e036878 =  5:  e33375f216 RF: update: Avoid overloading a variable for clarity
 6:  d88b6fb9c4 =  6:  a08d8878d4 OPT: update: Avoid repeatedly accessing Dataset.path in a loop
 7:  871fa9826c =  7:  d0698f56a0 RF: update: Break apart _update_repo() helper
 8:  37cfaa7701 =  8:  1d8dac50d8 ENH: update: Use same merge approach for plain and annex repos
 9:  7523a854f4 =  9:  e7dc6d2948 ENH: update: Avoid calling 'git pull' to merge changes
10:  2368b7377b = 10:  0ea6cf46da ENH: update: Don't set merge target to a guess that doesn't exist
11:  cf1987c7c8 ! 11:  7c46d0fa94 ENH: update: Try to fetch submodule commit explicitly if needed
    @@ datalad/distribution/tests/test_update.py
      
      """
      
    --
    --
     +import logging
      import os
      import os.path as op
    - from os.path import join as opj, exists
    + from os.path import (
     @@
    - from datalad.utils import knows_annex
    - from datalad.utils import rmtree
    - from datalad.utils import chpwd
    -+from datalad.utils import Path
    -+from datalad.utils import swallow_logs
    +     knows_annex,
    +     rmtree,
    +     chpwd,
    ++    Path,
    ++    swallow_logs,
    + )
      from datalad.support.gitrepo import (
          GitRepo,
    -     GitCommandError,
    +@@
    +     create_tree,
    +     ok_file_has_content,
    +     ok_clean_git,
    ++    assert_repo_status,
    +     assert_status,
    +     assert_result_count,
    +     assert_in_results,
    ++    assert_not_in_results,
    ++    SkipTest,
    +     slow,
    +     known_failure_windows,
      )
    - from datalad.support.annexrepo import AnnexRepo
    - 
    -+from nose import SkipTest
    - from nose.tools import eq_, assert_false, assert_is_instance, ok_
    - from datalad.tests.utils import with_tempfile, assert_in, \
    -     with_testrepos, assert_not_in
    - from datalad.tests.utils import create_tree
    - from datalad.tests.utils import ok_file_has_content
    - from datalad.tests.utils import ok_clean_git
    -+from datalad.tests.utils import assert_repo_status
    - from datalad.tests.utils import assert_status
    - from datalad.tests.utils import assert_result_count
    - from datalad.tests.utils import assert_in_results
    -+from datalad.tests.utils import assert_not_in_results
    - from datalad.tests.utils import slow
    - from datalad.tests.utils import known_failure_windows
    - 
     @@ datalad/distribution/tests/test_update.py: def test_multiway_merge(path):
          assert_status('ok', ds.update())
          # ATM we do not support multi-way merges
    @@ datalad/distribution/tests/test_update.py: def test_multiway_merge(path):
     +    ds_clone_s1.repo.call_git(["branch", "--unset-upstream"])
     +    ds_clone_s1.config.reload(force=True)
     +    ds_clone_s1.repo.call_git(["remote", "add", "other", ds_src_s1.path])
    ++    res = ds_clone.update(recursive=True, on_failure="ignore")
     +    if on_adjusted:
     +        assert_not_in_results(res, status="error")
     +    else:
    @@ datalad/distribution/tests/test_update.py: def test_multiway_merge(path):
     +        # it does if the revision wasn't brought down in the `fetch(all_=True)`
     +        # call.
     +        assert_in_results(
    -+            ds_clone.update(recursive=True, on_failure="ignore"),
    ++            res,
     +            status="impossible",
     +            path=ds_clone_s1.path,
     +            action="update")
    @@ datalad/distribution/update.py
          EnsureNone,
      )
      from datalad.support.annexrepo import AnnexRepo
    -+from datalad.support.gitrepo import (
    -+    GitCommandError,
    ++from datalad.support.exceptions import (
    ++    CommandError,
     +)
      from datalad.support.param import Parameter
      from datalad.interface.common_opts import (
    @@ datalad/distribution/update.py: def __call__(
     +                    try:
     +                        lgr.debug("Fetching revision %s directly for %s",
     +                                  revision, repo)
    -+                        repo.fetch(remote=sibling_, refspec=revision)
    -+                    except GitCommandError as exc:
    ++                        repo.fetch(remote=sibling_, refspec=revision,
    ++                                   git_options=["--recurse-submodules=no"])
    ++                    except CommandError as exc:
     +                        yield dict(
     +                            res,
     +                            status="impossible",
    @@ datalad/distribution/update.py: def __call__(
     -                    repo, curr_branch,
     -                    sibling_, tracking_remote)
     +                if needed_direct_fetch:
    -+                    lgr.debug("Needed to explicitly fetch revision %s in %s "
    ++                    lgr.debug("Needed to explicitly fetch revision %s in %s. "
     +                              "Using revision as merge target",
     +                              revision, repo)
     +                    merge_target = revision
12:  836ae37fb2 = 12:  05adce76f6 ENH: update: Prefer merge of submodule revision to a ref without it
```

</details>

kyleam commented:
Closing in favor of gh-4167.
